mindone/models/lora.py

```diff
@@ -0,0 +1,342 @@
+import logging
+from typing import List
+
+import mindspore as ms
+import mindspore.common.initializer as init
+from mindspore import nn
+from mindspore.nn.cell import Cell
+from mindspore.nn.layer.activation import get_activation
+from mindspore.ops.primitive import Primitive
+
+from ..utils.config import get_obj_from_str
+from ..utils.version_control import is_old_ms_version
+
+__all__ = [
+    "LoRADenseLayer",
+    "inject_trainable_lora",
+    "make_only_lora_params_trainable",
+    "merge_lora_to_model_weights",
+    "get_lora_params",
+]
+
+_logger = logging.getLogger(__name__)
+
+
+class LoRADenseLayer(nn.Cell):
+    """
+    Dense layer with lora injection, used to replace nn.Dense for lora fintuning.
+    """
+
+    def __init__(
+        self,
+        in_features,
+        out_features,
+        has_bias=True,
+        rank=4,
+        dropout_p=0.0,
+        scale=1.0,
+        dtype=ms.float32,
+        activation=None,
+    ):
+        super().__init__()
+
+        if rank > min(in_features, out_features):
+            raise ValueError(f"LoRA rank {rank} must be less or equal than {min(in_features, out_features)}")
+        self.rank = rank
+        self.scale = scale
+        self.dtype = dtype
+
```

Comment:
There are only dense layers in LoRA. Because we are still tuning the precision configuration for acceptable training results and performance, mannual precision setting for specific modules serves this purpose and may be dropped in future